src/puppetlabs/puppetdb/helpers.clj

```diff
@@ -0,0 +1,89 @@
+(ns puppetlabs.puppetdb.helpers
+  (:require [clojure.java.io :as io]
+            [clojure.tools.logging :as log]
+            [compojure.core :as compojure]
+            [puppetlabs.kitchensink.core :as kitchensink]
+            [puppetlabs.puppetdb.archive :as archive]
+            [puppetlabs.puppetdb.cheshire :as json]
+            [puppetlabs.puppetdb.cli.export :as export]
+            [puppetlabs.puppetdb.cli.import :as import]
+            [puppetlabs.puppetdb.http :as http]
+            [puppetlabs.puppetdb.middleware :as mid]
+            [puppetlabs.trapperkeeper.core :as trapperkeeper]
+            [ring.middleware.multipart-params :as mp]
+            [ring.util.io :as rio]))
+
+(def api-version :v4)
+
+(defn import-app
+  [infile command-fn]
+  (import/import! infile (partial import/process-tar-entry
+                                  command-fn)))
+
+(defn completion-wrapper
+  [query-fn child-href]
+  (let [[parent-str identifier child-str] (take-last 3 (clojure.string/split child-href #"/"))
+        entity (case child-str
+                 "metrics" :report-metrics
+                 "logs" :report-logs
+                 (keyword child-str))
+        query (case parent-str
+                "reports" (case child-str
+                            ("metrics" "logs") ["=" "hash" identifier]
+                            ["=" "report" identifier])
+                ("factsets" "catalogs") ["=" "certname" identifier])]
+    (query-fn entity api-version query nil doall)))
+
+(defn query-wrapper
+  [query-fn entity-str certname]
+  (let [entity (keyword entity-str)
+        query ["=" "certname" certname]]
+    (query-fn entity api-version query nil doall)))
+
+(def munge-nodes-data
+  (comp (partial kitchensink/mapvals str)
+        #(select-keys % [:certname :facts_timestamp :catalog_timestamp :report_timestamp])))
+
+(defn export-app
+  [outfile query-fn]
+  (export/export! outfile
+                  (->> (query-fn :nodes api-version nil nil doall)
+                       (map munge-nodes-data)
+                       (mapcat (partial export/get-node-data
+                                        (partial query-wrapper query-fn)
+                                        (partial completion-wrapper query-fn))))))
+
+(defn build-app
+  [submit-command-fn query-fn {:keys [backups-dir authorizer]}]
+  (-> (compojure/routes
+       ;; TODO Handle the following with the wrap-multipart-params
+       ;; TODO (mid/verify-content-type ["application/octet-stream"])
+       ;; TODO (mid/validate-query-params {:required ["archive"]})
```

Comment:
I made a ticket about this @wkalt, I know you don't like TODO comments but I figured it was best since this endpoint really doesn't have the right kind of validation we'll want eventually.